components/script/dom/fontface.rs

```diff
@@ -0,0 +1,568 @@
+
+        let sources = parsed_font_face_rule
+            .sources
+            .clone()
+            .expect("Sources should be non-None after validation");
+
+        // Let font face be a fresh FontFace object.
+        Self {
+            reflector: Reflector::new(),
+
+            // Set font face’s status attribute to "unloaded".
+            status: Cell::new(FontFaceLoadStatus::Unloaded),
+
+            // Set font face’s corresponding attributes to the serialization of the parsed values.
+            descriptors: DomRefCell::new(serialize_parsed_descriptors(parsed_font_face_rule)),
+
+            font_face_set: MutNullableDom::default(),
+            family_name: DomRefCell::new(family_name.clone()),
+            urls: DomRefCell::new(Some(sources)),
+            template: RefCell::default(),
+            font_status_promise,
+        }
+    }
+
+    pub(crate) fn new(
+        global: &GlobalScope,
+        proto: Option<HandleObject>,
+        font_family: DOMString,
+        source: StringOrArrayBufferViewOrArrayBuffer,
+        descriptors: &FontFaceDescriptors,
+        can_gc: CanGc,
+    ) -> DomRoot<Self> {
+        reflect_dom_object_with_proto(
+            Box::new(Self::new_inherited(
+                global,
+                font_family,
+                source,
+                descriptors,
+                can_gc,
+            )),
+            global,
+            proto,
+            can_gc,
+        )
+    }
+
+    pub(super) fn set_associated_font_face_set(&self, font_face_set: &FontFaceSet) {
+        self.font_face_set.set(Some(font_face_set));
+    }
+
+    pub(super) fn loaded(&self) -> bool {
+        self.status.get() == FontFaceLoadStatus::Loaded
+    }
+
+    pub(super) fn template(&self) -> Option<(LowercaseFontFamilyName, FontTemplate)> {
+        self.template.borrow().clone()
+    }
+}
+
+/// Implements the body of the setter for the descriptor attributes of the [`FontFace`] interface.
+///
+/// <https://drafts.csswg.org/css-font-loading/#fontface-interface>:
+/// On setting, parse the string according to the grammar for the corresponding @font-face
+/// descriptor. If it does not match the grammar, throw a SyntaxError; otherwise, set the attribute
+/// to the serialization of the parsed value.
+macro_rules! implement_setter_for_descriptor(
+    ($self:ident, $field: ident, $value: ident) => ({
+        let mut new_descriptors = $self.descriptors.borrow().clone();
+        new_descriptors.$field = $value;
+        let global = $self.global();
+        let parse_result = parse_font_face_descriptors(
+                &global,
+                &$self.family_name.borrow(),
+                None,
+                &new_descriptors
+        );
+        if let Some(ref parsed_font_face_rule) = parse_result {
+            *$self.descriptors.borrow_mut() = serialize_parsed_descriptors(parsed_font_face_rule);
+            Ok(())
+        } else {
+            Err(Error::Syntax)
+        }
```

Comment:
Could this part of the macro be turned into a helper method? Then each setter would just need to make new descriptor, set the field, and call the helper. That would allow more code reuse and avoid the need for metaprogramming, I think.